rust/src/statistics/storage/mod.rs

```diff
@@ -0,0 +1,633 @@
+// Copyright (c) 2025 Sonic Operations Ltd
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at soniclabs.com/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use of
+// this software will be governed by the GNU Lesser General Public License v3.
+
+use std::{
+    fmt,
+    io::Write,
+    path::{Path, PathBuf},
+};
+
+use dashmap::{DashMap, DashSet};
+
+use crate::{
+    error::BTResult,
+    storage::{self, Checkpointable, RootIdProvider, Storage},
+    sync::{Arc, Mutex, atomic, thread},
+    types::TreeId,
+};
+
+/// An operation that can be performed on the storage.
+#[derive(Hash, Eq, PartialEq, Debug, Copy, Clone)]
+enum StorageOperation {
+    Get,
+    Set,
+    Reserve,
+    Delete,
+}
+
+impl std::fmt::Display for StorageOperation {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            StorageOperation::Get => write!(f, "G"),
+            StorageOperation::Set => write!(f, "S"),
+            StorageOperation::Reserve => write!(f, "R"),
+            StorageOperation::Delete => write!(f, "D"),
+        }
+    }
+}
+
+/// A storage operation with the occurred timestamp and file offset that it refers to..
+#[derive(Hash, Eq, PartialEq, Debug, Copy, Clone)]
+struct StorageOpWithMetadata {
+    op: StorageOperation,
+    timestamp: u128,
+    offset: u64,
+}
+
+impl fmt::Display for StorageOpWithMetadata {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{},{},{}", self.op, self.timestamp, self.offset)
+    }
+}
+
+/// A storage wrapper that collects statistics about storage operations.
+/// It records the type of operation, timestamp, and offset of each operation.
+/// A background worker periodically flushes the collected statistics to disk.
+pub struct StorageStatistics<S: Storage> {
+    storage: S,
+    op_with_time_stats: Arc<DashSet<StorageOpWithMetadata>>,
+    _worker: StorageStatisticsWorker,
+}
+
+impl<S: Storage> StorageStatistics<S> {
+    /// Create a new storage statistics wrapper around the given storage.
+    pub fn try_new(storage: S, path: &Path) -> BTResult<Self, storage::Error> {
+        let op_with_time_stats = Arc::new(DashSet::new());
+        let _worker = StorageStatisticsWorker::new(op_with_time_stats.clone(), path)?;
+        Ok(Self {
+            storage,
+            op_with_time_stats,
+            _worker,
+        })
+    }
+
+    /// Log a storage operation with the current timestamp and offset.
+    fn add(&self, op: StorageOperation, offset: u64) {
+        let timestamp = std::time::SystemTime::now()
+            .duration_since(std::time::UNIX_EPOCH)
+            .expect("Time went backwards")
+            .as_micros();
```

Comment:
If we are only interested in the order of operations, an atomic that we increment on each op might be better suited.